.ci/test.sh

```diff
@@ -114,8 +128,23 @@ if [[ $TASK == "swig" ]]; then
     exit 0
 fi
 
-conda install -q -y -n $CONDA_ENV cloudpickle "dask=2021.9.1" "distributed=2021.9.1" joblib matplotlib numpy pandas psutil pytest scikit-learn scipy
-pip install graphviz  # python-graphviz from Anaconda is not allowed to be installed with Python 3.9
+
+${CONDA_INSTALL} -q -y -n $CONDA_ENV \
+    cloudpickle \
+    dask \
+    distributed \
+    joblib \
+    matplotlib \
+    numpy \
+    pandas \
+    psutil \
+    pytest \
+    scikit-learn \
+    scipy || exit -1
+
+# python-graphviz has to be installed separately to prevent conda from downgrading to pypy
+${CONDA_INSTALL} -q -y -n $CONDA_ENV \
+    python-graphviz || exit -1
```

Comment:
Only for the specific of case of `Linux sdist` (Python 3.7, ubuntu:14.04, gcc), I found that `conda` was downgrading Python from a cpython-based Python 3.7.12 to a PyPy-based Python 3.7.12: https://github.com/microsoft/LightGBM/pull/4953#issuecomment-1014707402

Was able to reproduce this locally in Docker.

```shell
docker run \
    --rm \
    --env CONDA=/opt/conda \
    --entrypoint="" \
    -it lightgbm/vsts-agent:ubuntu-14.04 \
    /bin/bash

# then run the following in the container
export PATH=${CONDA}/bin:${PATH}

conda create \
    --name test-env \
    -c conda-forge \
    --override-channels \
        python=3.7

conda config --set always_yes yes --set changeps1 no
conda update -q -y conda

source activate test-env
conda install -c conda-forge --override-channels -q -y -n test-env \
    cloudpickle \
    dask \
    distributed \
    joblib \
    matplotlib \
    numpy \
    pandas \
    psutil \
    python-graphviz \
    pytest \
    scikit-learn \
    scipy
```

This resulted in the following from the `conda install` logs

```text
The following packages will be DOWNGRADED:

  python                        3.7.12-hf930737_100_cpython --> 3.7.12-0_73_pypy
  python_abi                                    3.7-2_cp37m --> 3.7-2_pypy37_pp73
  setuptools                          60.5.0-py37h89c1867_0 --> 59.8.0-py37h9c2f6ca_0
```

So I started thinking "ok, what is changing in this PR? One thing I'm proposing changing in this PR is installing `python-graphviz` with conda instead of as a separate `pip install`, to reduce the risk of conflicts.

Funny enough, it seems like that change was the cause for `conda`  choosing PyPy. Using the sample code above, I found that installing `python-graphviz` by itself, separately, does not result in `conda` choosing to downgrade from a cpython-based Python to the PyPy one.